src/test/scala/chiselTests/PrefixSpec.scala

```diff
@@ -0,0 +1,278 @@
+        val addr = Input(UInt(10.W))
+        val dataIn = Input(UInt(8.W))
+        val dataOut = Output(UInt(8.W))
+      })
+
+      val smem = withModulePrefix("Foo") {
+        SyncReadMem(1024, UInt(8.W))
+      }
+
+      val cmem = withModulePrefix("Bar") {
+        Mem(1024, UInt(8.W))
+      }
+
+      val sram = withModulePrefix("Baz") {
+        SRAM(1024, UInt(8.W), 1, 1, 0)
+      }
+
+      smem.write(io.addr, io.dataIn)
+      io.dataOut := smem.read(io.addr, io.enable)
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+
+    val lines = """
+      {
+        "class":"chisel3.ModulePrefixAnnotation",
+        "target":"~Top|Top>smem",
+        "prefix":"Foo_"
+      },
+      {
+        "class":"chisel3.ModulePrefixAnnotation",
+        "target":"~Top|Top>cmem",
+        "prefix":"Bar_"
+      },
+      {
+        "class":"chisel3.ModulePrefixAnnotation",
+        "target":"~Top|Top>sram_sram",
+        "prefix":"Baz_"
+      }
+        """.linesIterator.map(_.trim).toSeq
+
+    matchesAndOmits(chirrtl)(lines: _*)()
+  }
+
+  it should "Definitions that appear within withModulePrefix get prefixed" in {
+    class Top extends Module {
+      val dfn = withModulePrefix("Foo") {
+        Definition(new AddOne(8))
+      }
+
+      val addone = Instance(dfn)
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+
+    val lines = """
+  module Foo_AddOne
+  module Top
+        """.linesIterator.map(_.trim).toSeq
+
+    matchesAndOmits(chirrtl)(lines: _*)()
+  }
+
+  it should "allow definitions to be instantiated within a withModulePrefix block without prefixing it" in {
+    class Child(defn: Definition[AddOne]) extends Module {
+      val addone = Instance(defn)
+    }
+
+    class Top extends Module {
+      val defn = Definition(new AddOne(8))
+
+      val child = withModulePrefix("Foo") {
+        Module(new Child(defn))
+      }
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+
+    val lines = """
+    module AddOne
+    module Foo_Child
+    public module Top
+        """.linesIterator.map(_.trim).toSeq
+  }
+
+  it should "withModulePrefix does not automatically affect ExtModules" in {
+    class Sub extends ExtModule {}
+
+    class Top extends Module {
+      val sub_foo = withModulePrefix("Foo") { Module(new Sub) }
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+
+    val lines = """
+      extmodule Sub
+        defname = Sub
+      module Top
+        """.linesIterator.map(_.trim).toSeq
+
+    matchesAndOmits(chirrtl)(lines: _*)()
+  }
+
+  it should "Using currentModulePrefix to force the name of an extmodule" in {
+    class Sub extends ExtModule {
+      override def desiredName = modulePrefix + "Sub"
+    }
+
+    class Top extends Module {
+      val sub_foo = withModulePrefix("Foo") { Module(new Sub) }
+    }
+
+    val chirrtl = emitCHIRRTL(new Top)
+
+    val lines = """
+      extmodule Foo_Sub
+        defname = Foo_Sub
+      module Top
+        """.linesIterator.map(_.trim).toSeq
```

Comment:
```suggestion
    val lines =
      """|extmodule Foo_Sub
         | defname = Foo_Sub
         |module Top
         |""".stripMargin.linesIterator.toSeq
```
This is a tiny nit, but the block comments in this file are somewhat inconsistently spaced. If you use `|` (and corresponding `.stripMargin`), then scalafmt will line them up like this for you (make sure to start the `"""|` on the following line, it lines up the `|` with each other). Then you also don't need the `map(_.trim)` then.